gateway/ln-gateway/src/client.rs

```diff
@@ -57,49 +64,83 @@ impl GatewayClientBuilder {
             mint_channel_id,
             gateway: gateway.clone(),
         });
-        registry.attach(GatewayClientInitV2 { gateway });
-
-        let db_path = self.work_dir.join(format!("{federation_id}.db"));
-
-        let rocksdb = fedimint_rocksdb::RocksDb::open(db_path.clone()).map_err(|e| {
-            GatewayError::DatabaseError(anyhow::anyhow!("Error opening rocksdb: {e:?}"))
-        })?;
-        let db = Database::new(rocksdb, ModuleDecoderRegistry::default());
+        registry.attach(GatewayClientInitV2 {
+            gateway: gateway.clone(),
+        });
 
         let mut client_builder = Client::builder(db)
             .await
             .map_err(GatewayError::DatabaseError)?;
         client_builder.with_module_inits(registry);
         client_builder.with_primary_module(self.primary_module);
         client_builder.with_connector(connector);
+        Ok(client_builder)
+    }
+
+    pub async fn build(
+        &self,
+        config: FederationConfig,
+        gateway: Arc<Gateway>,
+        mnemonic: &Mnemonic,
+    ) -> Result<fedimint_client::ClientHandleArc> {
+        let invite_code = config.invite_code.clone();
+        let federation_id = invite_code.federation_id();
+        let db_path = self.work_dir.join(format!("{federation_id}.db"));
 
-        let client_secret = if let Ok(secret) =
-            Client::load_decodable_client_secret::<[u8; 64]>(client_builder.db_no_decoders()).await
-        {
-            secret
+        let (db, root_secret) = if db_path.exists() {
+            let rocksdb = fedimint_rocksdb::RocksDb::open(db_path.clone()).map_err(|e| {
+                GatewayError::DatabaseError(anyhow::anyhow!("Error opening rocksdb: {e:?}"))
+            })?;
+            let db = Database::new(rocksdb, ModuleDecoderRegistry::default());
+            let root_secret = self.client_plainrootsecret(&db).await?;
+            (db, root_secret)
         } else {
-            info!("Generating secret and writing to client storage");
-            let secret = PlainRootSecretStrategy::random(&mut thread_rng());
-            Client::store_encodable_client_secret(client_builder.db_no_decoders(), secret)
-                .await
-                .map_err(GatewayError::ClientStateMachineError)?;
-            secret
+            let db = gateway
+                .gateway_db
+                .with_prefix(federation_id.to_prefix().to_bytes());
```

Comment:
Use `mint_channel_id` here, also add check to make sure we don't re-use a database for multiple clients

Reply:
Done. Also added a check that verifies the stored `ClientConfig`'s federation ID matches the federation ID we are creating. This should catch any accidental re-use of databases.